pages/api/VolunteerRoleService.ts

```diff
@@ -0,0 +1,187 @@
+      tableId || process.env.NEXT_PUBLIC_AIRTABLE_TABLE_VOLUNTEER_ROLES
+    )
+    this.toolsWeUseTable = this.base(process.env.NEXT_PUBLIC_AIRTABLE_TABLE_TOOLS_WE_USE)
+  }
+
+  async getAllActiveRoles(): Promise<DASVolunteerRoleBasicInfo[]> {
+    try {
+      const records = await this.volunteerRolesTable
+        .select({
+          view: 'Grid view',
+        })
+        .all()
+
+      const activeRoles = records.filter(
+        (record) => record.get('Status') === 'Active'
+      )
+
+      const volunteerRolesData = activeRoles.map((record) => {
+        return {
+          id: record.id,
+          role: String(record.fields.Role),
+          key: String(record.fields.Key),
+        }
+      })
+
+      return volunteerRolesData
+    } catch (err) {
+      console.log(err)
+      throw err
+    }
+  }
+  async getRoleDetailsByName(
+    roleKey: string
+  ): Promise<DASVolunteerRole | null> {
+    try {
+      const recordData = await this.volunteerRolesTable
+        .select({
+          view: 'Grid view',
+          maxRecords: 1,
+          filterByFormula: `{Key} = '${roleKey}'`,
+        })
+        .firstPage()
+      if (recordData.length === 0) {
+        return null
+      }
+      const record = recordData[0]
+
+      if (record.fields.Status !== 'Active') {
+        return null
+      }
+      let keyTechnologies = null
+
+      if (record.fields['Key technologies'] && Array.isArray(record.fields['Key technologies'])) {
+        const technologiesRecordIds:Array<any> = record.fields['Key technologies']
+        try {
+          const technologiesData = await this.toolsWeUseTable
+            .select({
+              view: 'Grid view',
+            })
+            .all()
+          const roleTechnologies = technologiesData.filter((tech) => {
+            return technologiesRecordIds.includes(tech.id)
+          })
+          const technologiesMap = new Map()
+          roleTechnologies.forEach((tech) => {
+            const category = tech.fields?.Category || 'Other'
+            const name = tech.fields?.['Name']
+            if (!technologiesMap.has(category)) {
+              technologiesMap.set(category, [])
+            }
+            technologiesMap.get(category).push(name)
+          })
+          keyTechnologies = []
+          technologiesMap.forEach((value, key) => {
+            keyTechnologies.push(`${key}: ${value.join(', ')}`)
+          })
+          keyTechnologies = keyTechnologies.sort((a, b) => {
+            if (a.startsWith('Other')) {
+              return 1
+            } else if (b.startsWith('Other')) {
+              return -1
+            } else {
+            }
+          })
+        } catch (error) {}
+      }
+      const volunteerRoleData: DASVolunteerRole = {
+        role:
+          typeof record.fields.Role === 'string' ? record.fields.Role : null,
+        key: typeof record.fields.Key,
+        location:
+          typeof record.fields?.Location === 'string'
```

Comment:
let's look for a way to get rid of this copy&paste.

Reply:
Maybe something like this would help.  The methods could be extracted into a helper class and perhaps applied to other future airtable requests.

  ```
getStringFieldRecord(record: Record<FieldSet>, field: string): string {
    return record.fields ? typeof record.fields[field] === 'string'
      ? record.fields[field] as string
      : null : null
  }

  getNumberFieldRecord(record: Record<FieldSet>, field: string): number {
    return record.fields ? typeof record.fields[field] === 'number'
      ? record.fields[field] as number
      : null : null
  }

  getSplitFieldRecord(record: Record<FieldSet>, field: string): string[] {
    return record.fields ? typeof record.fields[field] === 'string'
      ? (record.fields[field] as string).split('\n')
      : null : null
  }
....
        preferredQualifications: this.getStringFieldRecord(record, 'Preferred Qualifications'),
        keyAttributesToSuccess: this.getSplitFieldRecord(record, 'Key attributes for success'),
....
```

Reply:
![javascript](https://github.com/openseattle/open-seattle-website/assets/11780107/dfc76f2f-047c-4681-acb8-10bd06f1d41b)

It's still Javascript where objects are maps